File: C++/Basic Hashing/count_frequencies.cpp
#include <bits/stdc++.h>

using namespace std;

/*
    Solution for the problem: https://www.codingninjas.com/studio/problems/count-frequency-in-a-range_8365446
    Array Hashing approach
    Time Complexity = O(n)
    Space Complexity = O(n)
*/

vector<int> countFrequency(int n, int x, vector<int> &nums)
{
    vector<int> frequencies(n, 0);
    for (auto i : nums)
        if (i <= n)
            frequencies[i - 1]++;

    return frequencies;
}

int main()
{
    int n, x, buffer;
    vector<int> nums, frequencies;
    cout << "Enter the size of input array: " << endl;
    cin >> n;
    cout << "Enter the max range of values in the array (1 - max): " << endl;
    cin >> x;
    cout << "Enter the values for the array: " << endl;
    for (int i = 0; i < n; i++)
    {
        cin >> buffer;
        if (buffer <= x && buffer >= 1)
            nums.push_back(buffer);
    }
    cout << "Input array: " << endl;
    for (auto i : nums)
        cout << i << " ";
    frequencies = countFrequency(n, x, nums);
    cout << "\nFrequency of, " << endl;
    for (int i = 1; i <= n; i++)
        cout << i << " -> " << frequencies[i - 1] << endl;

    return 0;
}

File: Python/Basic Hashing/count_frequencies.py
'''
    Solution for the problem: https://www.codingninjas.com/studio/problems/count-frequency-in-a-range_8365446
    Array Hashing approach
    Time Complexity = O(n)
    Space Complexity = O(n)
'''


def process_list_input(i):
    if i.isdigit() and int(i) < x:
        return int(i)
    else:
        return 0


def countFrequency(n, x, nums):
    frequencies = [0] * n
    for i in nums:
        if i < n:
            frequencies[i - 1] += 1

    return frequencies


print("Enter the size of input array: ")
n = int(input())
print("Enter the max range of values in the array (1 - max): ")
x = int(input())
print("Enter the values for the array: ")
nums = list(map(process_list_input, input().split()))
nums = [i for i in nums if i]
nums = nums[:n]
print("Input array:", nums)
while len(nums) < n:
    print("Re-enter the values for the array within the max range: ")
    nums = list(map(process_list_input, input().split()))
    nums = [i for i in nums if i]
    nums = nums[:n]
    print("Input array:", nums)
frequencies = countFrequency(n, x, nums)
print("Frequency of, ")
for i in range(1, n + 1):
    print(i, "->", frequencies[i - 1])
